============================================
The ``7. Lazy matrix multiplication`` module
============================================

>>> lazy_matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul


>>> lazy_matrix_mul([[1, 2], [3, 4], [3, 4]], [[5, 6, 1], [7, 8, 2]])
matrix([[19, 22,  5],
        [43, 50, 11],
        [43, 50, 11]])


>>> lazy_matrix_mul([[1, 2, 3], [3, 4, 5]], [[1, 2], [3, 4]])
Traceback (most recent call last):
	...
ValueError: shapes (2,3) and (2,2) not aligned: 3 (dim 1) != 2 (dim 0)


>>> lazy_matrix_mul([[3]])
Traceback (most recent call last):
	...
TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'


>>> lazy_matrix_mul()
Traceback (most recent call last):
	...
TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'


>>> lazy_matrix_mul([], [[3]])
Traceback (most recent call last):
	...
ValueError: shapes (1,0) and (1,1) not aligned: 0 (dim 1) != 1 (dim 0)


>>> lazy_matrix_mul([[]], [[3]])
Traceback (most recent call last):
	...
ValueError: shapes (1,0) and (1,1) not aligned: 0 (dim 1) != 1 (dim 0)


>>> lazy_matrix_mul("isa", [[9]])
Traceback (most recent call last):
	...
TypeError: Scalar operands are not allowed, use '*' instead


>>> lazy_matrix_mul([[8]], "arzu")
Traceback (most recent call last):
	...
TypeError: Scalar operands are not allowed, use '*' instead


>>> lazy_matrix_mul([1, 3, 4], [[9]])
Traceback (most recent call last):
	...
TypeError: object of type 'int' has no len()

